consensus key rotation (#727)
aptos-labs/developer-docs: apps/nextra/pages/en/network/nodes/validator-node/modify-nodes/rotate-consensus-key.mdx

---
title: "Rotate consensus key"
---

import { Callout, Steps } from 'nextra/components';

# Rotate consensus key

Consensus key rotation is an action taken by a node operator to change the identity of a validator they control.
It happens typically when the corresponding private key is (potentially) lost/leaked,

> or every X months as a common security practice.

Reviewer: ```suggestion
or every few months as a common security practice.
```

Author: fixed

Below is the step-by-step guide of performing a consensus key rotation with detailed examples.

## New key generation

> Using the following [aptos CLI](https://aptos.dev/en/build/cli) command to generate a new consensus identity.

Reviewer: ```suggestion
Using the following [Aptos CLI](https://aptos.dev/en/build/cli) command to generate a new consensus identity.
```

Author: fixed

In this example, identity files are saved to directory `/new/key/root`.

> ```bash
> aptos genesis generate-keys --output-dir /new/key/root
> ```

Reviewer: ```suggestion
```bash filename="Terminal"
aptos genesis generate-keys --output-dir /new/key/root
```
```

Author: fixed

## Add new private key to node

Edit the node config yaml as described below.
```yaml
# ...
consensus:
    # ...
    safety_rules:
        # ...
        initial_safety_rules_config:
            from_file:
                # ...
                identity_blob_path: /old/key/root/validator-identity.yaml
                overriding_identity_paths:              # new!
                - /new/key/root/validator-identity.yaml # new!
```
<Callout type="info">
Don't delete `identity_blob_path`. It is a required field.

Field `overriding_identity_paths` is optional, and allows multiple identities to be specified.

> 

Reviewer: Oof this is messy, no wonder there's a guide

> Don't delete `identity_blob_path`. It is a required field.
> 
> Field `overriding_identity_paths` is optional, and allows multiple identities to be specified.
> 
> Among all the identities, the node finds the one that appears on-chain validator set, and uses it as the consensus key.

Reviewer: This blurb is confusing.

Maybe...
```suggestion
* Field `identity_blob_path` is required.
* Field `overriding_identity_paths` is optional, and allows multiple identities to be specified for rotation purposes.
* The node will choose the identity that matches the on-chain state.
```

Author: fixed

</Callout>

<Callout type="info">
Old keys may still be available even if you update `identity_blob_path`
or delete items from `overriding_identity_paths`,
due to the reasons below.

There is a separate **secure storage** (typically a file `secure_storage.json`) hosting all identities.
The content of `identity_blob_path` and `overriding_identity_paths` are added to the secure storage on node start, if missing.
However, keys are never deleted automatically from the storage.

To completely wipe out an old key, you will have to manually make changes to the secure storage.
See the end of the guide for more details.

> </Callout>

Reviewer: Why are there two callouts?

I would just remove the callouts altogether

Author: deleted this callout, and only kept the first one.

        hex:0x8e27fd9300433191b1123217928a6f5190a6ec344ea8623555712a850029b34f5c4bab68df7568b48bcced408cde5174064284407ee760df5dbf12d1c6090589ea1a692997018aca740e91d2182e5715c7745565fe99361e279ccfcfa10ae1f7
```

## Wait for the next epoch

The new key should become effective after the next epoch change (which happens every 2 hours).

## (Advanced) Old key clean-up

In the secure storage (typically a file named `secure_storage.json`),
consensus identities are organized as follows.
- Field `consensus` contains the default consensus identity.
  - Typically created when you start the node for the first time.
  - Must exist.
- Field `consensus_X` contains additional consensus identities with `X` as its public key.
  - Created because you once added it to `overriding_identity_paths` in node config yaml.

**Currently, identities won't be deleted automatically from secure storage**,
even if you delete the corresponding identity from the node config yaml.

If you need to ensure the old key is completely gone, manually clean-up is needed.

Here is an example of `secure_storage.json`.

> ```json
> {
>     // ...
>     "consensus": {
>         "data": "GetResponse",
>         "last_update": 1731372563,
>         "value": "0x221f6fbfefa0a40b84c88fbb546a0884977dcc56719a96ed4e5d69b6a4ff58c8"
>     },
>     "consensus_acb7859468ca85cf9935e64ebb2b9b3fa8187de42d541acebaf732365a0131eaa994098f9d0d7e6b8ddea8ef11e16c55": {
>         "data": "GetResponse",
>         "last_update": 1731383387,
>         "value": "0x18d5098b3819d5fb0fc208b8bb0946b263f961f60716fc4d10d4b010fdb89a55"
>     },
>     // ...
> }

Reviewer: To match the comment by graphite, which I agree for syntax highlighting
```suggestion
```js
{
    // ...
    "consensus": {
        "data": "GetResponse",
        "last_update": 1731372563,
        "value": "0x221f6fbfefa0a40b84c88fbb546a0884977dcc56719a96ed4e5d69b6a4ff58c8"
    },
    "consensus_acb7859468ca85cf9935e64ebb2b9b3fa8187de42d541acebaf732365a0131eaa994098f9d0d7e6b8ddea8ef11e16c55": {
        "data": "GetResponse",
        "last_update": 1731383387,
        "value": "0x18d5098b3819d5fb0fc208b8bb0946b263f961f60716fc4d10d4b010fdb89a55"
    },
    // ...
}
```

Author: fixec
